libvast/vast/arrow_extension_types.hpp

```diff
@@ -158,24 +173,350 @@ class pattern_extension_type : public arrow::ExtensionType {
+            get(x, sum_type_token<Ts, Indices>{}));
+        }...};
+    }
+    (types{}, std::make_integer_sequence<int, detail::tl_size<types>::value>());
+    const auto dispatch = table[index_from_type(*x.type())];
+    VAST_ASSERT(dispatch);
+    return dispatch(x, std::forward<Visitor>(v), std::forward<Args>(xs)...);
+  }
+};
+
+template <>
+struct sum_type_access<arrow::DataType> final {
+  using types = sum_type_access<arrow::Array>::data_types;
+
+  using type0 = detail::tl_head_t<types>;
+
+  static constexpr bool specialized = true;
+
+  template <class T, int Index>
+  static bool is(const arrow::DataType& x, sum_type_token<T, Index>) {
+    if (x.id() != T::type_id)
+      return false;
+    if constexpr (arrow::is_extension_type<T>::value)
+      return static_cast<const arrow::ExtensionType&>(x).extension_name()
+             == T::vast_id;
+    return true;
+  }
+
+  template <class T, int Index>
+  static const T& get(const arrow::DataType& x, sum_type_token<T, Index>) {
+    return static_cast<const T&>(x);
+  }
+
+  template <class T, int Index>
+  static T& get(arrow::DataType& x, sum_type_token<T, Index>) {
+    return static_cast<T&>(x);
+  }
+
+  template <class T, int Index>
+  static const T*
+  get_if(const arrow::DataType* x, sum_type_token<T, Index> token) {
+    if (x && is(*x, token))
+      return &get(*x, token);
+    return nullptr;
+  }
+
+  template <class T, int Index>
+  static T* get_if(arrow::DataType* x, sum_type_token<T, Index> token) {
+    if (x && is(*x, token))
+      return &get(*x, token);
+    return nullptr;
+  }
+
+  template <class Result, class Visitor, class... Args>
+  static auto apply(const arrow::DataType& x, Visitor&& v, Args&&... xs)
+    -> Result {
+    // A dispatch table that maps variant type index to dispatch function for
+    // the concrete type.
+    static constexpr auto table = []<class... Ts, int... Indices>(
+      detail::type_list<Ts...>,
+      std::integer_sequence<int, Indices...>) noexcept {
+      return std::array{
+        // decay to function pointer
+        +[](const arrow::DataType& x, Visitor&& v, Args&&... xs) -> Result {
+          auto xs_as_tuple = std::forward_as_tuple(xs...);
+          auto indices = detail::get_indices(xs_as_tuple);
+          return detail::apply_args_suffxied(
+            std::forward<decltype(v)>(v), std::move(indices), xs_as_tuple,
+            get(x, sum_type_token<Ts, Indices>{}));
+        }...};
+    }
+    (types{}, std::make_integer_sequence<int, detail::tl_size<types>::value>());
+    const auto dispatch
+      = table[sum_type_access<arrow::Array>::index_from_type(x)];
+    VAST_ASSERT(dispatch);
+    return dispatch(x, std::forward<Visitor>(v), std::forward<Args>(xs)...);
+  }
+};
+
+namespace detail {
+template <class Types>
+struct tl_map_shared_ptr;
+template <class... Ts>
+struct tl_map_shared_ptr<detail::type_list<Ts...>> {
+  using type = detail::type_list<std::shared_ptr<Ts>...>;
+};
```

Comment:
```suggestion
} // namespace caf

namespace vast::detail {

template <class Types>
struct tl_map_shared_ptr;

template <class... Ts>
struct tl_map_shared_ptr<detail::type_list<Ts...>> {
  using type = detail::type_list<std::shared_ptr<Ts>...>;
};

} // namespace vast::detail

namespace caf {
```

Either that, or put it into `vast/detail/type_traits.hpp` directly with a small comment above it. But don't put anything but a template specialization into another library's namespace.

Reply:
moved over to `vast/detail/type_traits.hpp` as suggested.